### PR TITLE
app/{vminsert,vmagent}: ignore series on exceeding `-maxLabelsPerTimeseries`, `-maxLabelValueLen` or `-maxLabelNameLen` flag limits

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
### Describe Your Changes

related issues:
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7661

### Checklist

The following checks are **mandatory**:

- [ ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

**Review by f41gh7 — CHANGES_REQUESTED** [review]
Thanks for the PR. Please see the comments.

 I strongly recommend do not mix huge pile of changes into the single PR.  It increases complexity and makes it hard to review. 
 It's better to move refactoring, docker-compose and dashboard related changes into separate follow-up PRs later.

 I have a note about proposed changes. 
 For me it seems naturally to drop timeseries as soon as possible. Before it even reach first relabeling rule. In this case, it reduces resource usage and allow to protect metrics receiver from excessive resource usage.

 It's still possible, that resulting relabeling rule could exceed limit, but it shouldn't be our concern. 

 TLDR:  I'd like to drop timeseries before streaming aggregation and global relabeling at `vmagent` side. 
 The same is for `vminsert` (and single node).

**Comment by f41gh7** [comment]
@hagen1778  @Haleygo  WDYT?

File: app/vminsert/common/insert_ctx.go [hunk 1]
```diff
@@ -51,7 +52,7 @@ func cleanMetricRow(mr *storage.MetricRow) {
 	mr.MetricNameRaw = nil
 }
 
-func (ctx *InsertCtx) marshalMetricNameRaw(prefix []byte, labels []prompb.Label) []byte {
+func (ctx *InsertCtx) marshalMetricNameRaw(prefix []byte, labels []prompbmarshal.Label) []byte {
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vminsert/common/insert_ctx.go:55]: this change already exists as a separate [PR](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7409), but without it here code from `prompbmarshal/limits` should be duplicated for `prompb` package

**Comment by AndrewChubatiuk** [comment]
@f41gh7 moved removal of exceeding series before relabeling. if any processing is needed users can put vmagent in front with a bigger limits than vminsert has to make transformation

**Comment by Haleygo** [comment]
>For me it seems naturally to drop timeseries as soon as possible. Before it even reach first relabeling rule. In this case, it reduces resource usage and allow to protect metrics receiver from excessive resource usage.

I disaggree, I think the goal of those limits is to protect VictoriaMetrics behind vmagent, we just need to ensure data get checked in vmagent before it reaches vminsert. 
If data is exceeds these limits, it might be hard for user(vmagent admin) to request request changes from the sources(app admin), the metric sources also can have multiple stages that add extra labels, some of which may be unnecessary. However, users can easily apply relabeling rules to remove meaningless labels or merge useful ones. 
And, users can consistently set the same limits on both vmagent and VictoriaMetrics.

**Comment by Haleygo** [comment]
> @f41gh7 moved removal of exceeding series before relabeling. if any processing is needed users can put vmagent in front with a bigger limits than vminsert has to make transformation

@AndrewChubatiuk as I can see, the check is still behind relabeling.

Another thing, how about adding those limits on job level as well, so user can have different limit for some scrape jobs over the global one?

File: app/vmagent/remotewrite/remotewrite.go [hunk 2]
```diff
@@ -485,6 +484,14 @@ func tryPush(at *auth.Token, wr *prompbmarshal.WriteRequest, forceDropSamplesOnF
 			rowsCountAfterRelabel := getRowsCount(tssBlock)
 			rowsDroppedByGlobalRelabel.Add(rowsCountBeforeRelabel - rowsCountAfterRelabel)
 		}
+		idxDiff := 0
+		for i, ts := range tssBlock {
+			if storagelimits.ExceedingLabels(ts.Labels) {
+				idxDiff++
+			} else if idxDiff > 0 {
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vmagent/remotewrite/remotewrite.go:491]: if idxDiff is greater than 0 all data in tssBlock will be shifted by idxDiff

[the remark this answers is not in the document]
Comment (Haleygo) [reply app/vmagent/remotewrite/remotewrite.go:491]: Even with the change is https://github.com/VictoriaMetrics/VictoriaMetrics/compare/66641e027317d68a6aedaf7513c04c4b2992b277..326720c16648e60de61917b7c437ee8685cc7efb, the code is still incorrect, as the index of wrong sample is lost(let's say tssBolck[2] exceeds the limit), and why is `tss` being updated as well?
```
		idxDiff := 0
		for i, ts := range tssBlock {
			if storagelimits.ExceedingLabels(ts.Labels) {
				idxDiff++
			} else if idxDiff > 0 {
				tss[i-idxDiff] = tss[i]
			}
		}
		tssBlock = tssBlock[:len(tssBlock) - idxDiff]
```

File: app/vminsert/common/insert_ctx.go [hunk 3]
```diff
@@ -60,15 +61,21 @@ func (ctx *InsertCtx) marshalMetricNameRaw(prefix []byte, labels []prompb.Label)
...
+	}
 	metricNameRaw := ctx.marshalMetricNameRaw(prefix, labels)
 	return ctx.addRow(metricNameRaw, timestamp, value)
 }
 
 // WriteDataPointExt writes (timestamp, value) with the given metricNameRaw and labels into ctx buffer.
 //
 // It returns metricNameRaw for the given labels if len(metricNameRaw) == 0.
-func (ctx *InsertCtx) WriteDataPointExt(metricNameRaw []byte, labels []prompb.Label, timestamp int64, value float64) ([]byte, error) {
+func (ctx *InsertCtx) WriteDataPointExt(metricNameRaw []byte, labels []prompbmarshal.Label, timestamp int64, value float64) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vminsert/common/insert_ctx.go:75]: almost each vminsert ingestion protocol counts all samples even if they are dropped later
[here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/datadogv1/request_handler.go#L46) we increment rowsLen and then it can be dropped [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/datadogv1/request_handler.go#L69) without any metric decrement

[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vminsert/common/insert_ctx.go:75]: also see a comment [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/native/request_handler.go#L38)

File: app/vmagent/remotewrite/remotewrite.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vmagent/remotewrite/remotewrite.go:491]: > and why is tss being updated as well

thanks, fixed it

> as the index of wrong sample is lost

we don't need it as we're just shifting all series by an amount of exceeding ones

File: app/vmagent/remotewrite/remotewrite.go [hunk 4]
```diff
@@ -485,6 +484,15 @@ func tryPush(at *auth.Token, wr *prompbmarshal.WriteRequest, forceDropSamplesOnF
...
+				tssBlock[i-idxDiff] = tssBlock[i]
+			}
+		}
+		tssBlock = tssBlock[:len(tssBlock)-idxDiff]
```
Comment (Haleygo) [inline app/vmagent/remotewrite/remotewrite.go:495]: It's better to check if len(tssBlock)==0 afterwards, and skip following steps if it is.

File: docs/vmagent.md [hunk 5]
```diff
@@ -1893,6 +1893,10 @@ See the docs at https://docs.victoriametrics.com/vmagent/ .
   -maxInsertRequestSize size
      The maximum size in bytes of a single Prometheus remote_write API request
      Supports the following optional suffixes for size values: KB, MB, GB, TB, KiB, MiB, GiB, TiB (default 33554432)
+  -maxLabelValueLen int
+     The maximum length of label values in the accepted time series. Series, with longer label values are dropped. In this case the `vm_series_dropped_total{reason="too_long_label_value"}` metric at /metrics page is incremented (default 4096).
```
Comment (Haleygo) [inline docs/vmagent.md:1897]: Please also mention the limits is applied on series after `-remoteWrite.relabelConfig`.

File: app/vmagent/remotewrite/remotewrite.go (hunk 4 above)
Comment (f41gh7) [reply app/vmagent/remotewrite/remotewrite.go:495]: Looks like more idiomatic way to filter in-place: 
```
# slice tricks https://go.dev/wiki/SliceTricks#filter-in-place
n := 0
for _, x := range a {
    if keep(x) {
        a[n] = x
        n++
    }
}
a = a[:n]
```

 At VictoriaMetrics code base more often tmpSlice pattern is used:
```
tmpA := a[:0]
for _, x := range a {
    if keep(x) {
        tmpA = append(tmpA,x)
    }
}
a = tmpA
```

File: app/vminsert/common/streamaggr.go [hunk 6]
```diff
@@ -266,6 +266,9 @@ func pushAggregateSeries(tss []prompbmarshal.TimeSeries) {
 			}
 			ctx.AddLabel(name, label.Value)
 		}
+		if ctx.AreLabelsInvalid() {
```
Comment (f41gh7) [inline app/vminsert/common/streamaggr.go:269]: It's impossible case. 

 More over, it actually proves that all this changes for renaming `prompb` to `prompbmarshal` import is not needed.

 Please rollback it.

Comment (AndrewChubatiuk) [reply app/vminsert/common/streamaggr.go]: done

File: app/vmagent/remotewrite/remotewrite.go [hunk 7]
```diff
@@ -485,6 +484,16 @@ func tryPush(at *auth.Token, wr *prompbmarshal.WriteRequest, forceDropSamplesOnF
 			rowsCountAfterRelabel := getRowsCount(tssBlock)
 			rowsDroppedByGlobalRelabel.Add(rowsCountBeforeRelabel - rowsCountAfterRelabel)
 		}
+		tssBlockTmp := tssBlock[:0]
+		for _, ts := range tssBlock {
+			if storagelimits.ExceedingLabels(ts.Labels) {
```
Comment (f41gh7) [inline app/vmagent/remotewrite/remotewrite.go:489]: I think condition should be inverted to `if !storagelimits.ExceedingLabels(ts.Labels)` if labels are not exceeding limits.

File: app/victoria-metrics/self_scraper.go [hunk 8]
```diff
@@ -48,7 +48,7 @@ func selfScraper(scrapeInterval time.Duration) {
 	var bb bytesutil.ByteBuffer
 	var rows prometheus.Rows
 	var mrs []storage.MetricRow
-	var labels []prompb.Label
+	var labels []prompbmarshal.Label
```
[the remark this answers is not in the document]
Comment (hagen1778) [reply app/victoria-metrics/self_scraper.go:51]: Agree, this should be a separate change. @AndrewChubatiuk could you remove it from this PR?

[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/victoria-metrics/self_scraper.go:51]: I cannot do it as it leads to same `ExceedingLabels` function implementations for prompb.Label and prompbmarshal.Label structures. [This PR](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7409) should be merged first

File: lib/storage/limits/limits.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+	// droppedSeriesWithTooLongLabelName is the number of dropped series which contain labels with too long names
+	droppedSeriesWithTooLongLabelName atomic.Uint64
+
+	// droppedSeriesWithTooLongLabelValue is the number of dropped series which contain labels with too long values
+	droppedSeriesWithTooLongLabelValue atomic.Uint64
+)
+
+var (
+	_ = metrics.NewGauge(`vm_series_dropped_total{reason="too_many_labels"}`, func() float64 {
```
Comment (hagen1778) [inline lib/storage/limits/limits.go:39]: wdyt about using `vm_rows_ignored_total` name instead to remain consistent with other similar metrics exposed by vmstorage?

File: lib/storage/limits/limits.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func trackDroppedSeriesWithTooLongLabelValue(l *prompbmarshal.Label, labels []prompbmarshal.Label) {
+	droppedSeriesWithTooLongLabelValue.Add(1)
+	select {
+	case <-droppedSeriesWithTooLongLabelValueLogTicker.C:
+		label := *l
+		// Do not call logger.WithThrottler() here, since this will result in increased CPU usage
+		// because prompbmarshal.LabelsToString() will be called with each trackDroppedSeriesWithTooLongLabelValue call.
+		logger.Warnf("drop series with a value %s for label %s because its length=%d exceeds -maxLabelValueLen=%d; "+
```
Comment (hagen1778) [inline lib/storage/limits/limits.go:70]:
```suggestion
		logger.Warnf("dropping series with %s=%q label for %s; label %s length=%d exceeds -maxLabelValueLen=%d; "+
```

File: lib/storage/limits/limits.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+			"original labels: %s; either reduce the label value length or increase -maxLabelValueLen command-line flag value",
```
Comment (hagen1778) [inline lib/storage/limits/limits.go:71]:
```suggestion
			"either reduce the label value length or increase -maxLabelValueLen command-line flag value",
```

File: lib/storage/limits/limits.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func trackDroppedSeriesWithTooLongLabelName(l *prompbmarshal.Label, labels []prompbmarshal.Label) {
+	droppedSeriesWithTooLongLabelName.Add(1)
+	select {
+	case <-droppedSeriesWithTooLongLabelNameLogTicker.C:
+		label := *l
+		// Do not call logger.WithThrottler() here, since this will result in increased CPU usage
+		// because prompbmarshal.LabelsToString() will be called with each trackDroppedSeriesWithTooLongLabelName call.
+		logger.Warnf("drop series with a value for label %s because its length=%d exceeds %d; "+
```
Comment (hagen1778) [inline lib/storage/limits/limits.go:84]:
```suggestion
		logger.Warnf("dropping series with label %q for %s; label name length=%d exceeds max allowed %d - consider reducing label name length.",
```

File: lib/storage/limits/limits.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+			"original labels: %s; consider reducing the label name length",
```
Comment (hagen1778) [inline lib/storage/limits/limits.go:85]:
```suggestion
```

File: deployment/docker/rules/alerts-cluster.yml [hunk 14]
```diff
@@ -141,13 +141,13 @@ groups:
             See also https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3976#issuecomment-1476883183"
 
       - alert: LabelsLimitExceededOnIngestion
-        expr: increase(vm_metrics_with_dropped_labels_total[5m]) > 0
+        expr: increase(vm_series_dropped_total{reason="too_many_labels"}[5m]) > 0
```
Comment (hagen1778) [inline deployment/docker/rules/alerts-cluster.yml:144]: If we use `vm_rows_ignored_total` metric name instead, we can drop these rules

File: deployment/docker/rules/alerts.yml [hunk 15]
```diff
@@ -121,13 +121,13 @@ groups:
             See also https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3976#issuecomment-1476883183"
 
       - alert: LabelsLimitExceededOnIngestion
-        expr: increase(vm_metrics_with_dropped_labels_total[5m]) > 0
+        expr: increase(vm_series_dropped_total{reason="too_many_labels"}[5m]) > 0
```
Comment (hagen1778) [inline deployment/docker/rules/alerts.yml:124]: If we use `vm_rows_ignored_total` metric name instead, we can drop these rules

File: deployment/docker/rules/alerts-health.yml [hunk 16]
```diff
@@ -132,11 +133,12 @@ groups:
             either reduce the size of label values or increase `-maxLabelValueLen`.
 
       - alert: TooLongLabelNames
-        expr: increase(vm_too_long_label_names_total[5m]) > 0
+        expr: increase(vm_series_dropped_total{reason="too_long_label_name"}[5m]) > 0
```
Comment (hagen1778) [inline deployment/docker/rules/alerts-health.yml:136]: If we use `vm_rows_ignored_total` metric name instead, we can drop these rules

File: docs/changelog/CHANGELOG.md [hunk 17]
```diff
@@ -23,11 +23,13 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
...
 * FEATURE: [stream aggregation](https://docs.victoriametrics.com/stream-aggregation/): add `ignore_first_sample_interval` param to [aggregation config](https://docs.victoriametrics.com/stream-aggregation/#stream-aggregation-config). It allows users to control the time interval when aggregation skips sending aggregated samples to avoid unexpected spikes in values. By default, this interval is set to x2 of `staleness_interval`. The new setting is applicable only to `total`, `total_prometheus`, `increase`, `increase_prometheus` and `histogram_bucket` outputs. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7116) for details. Thanks to @iyuroch for the [pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/7313).
+* FEATURE: [vmagent](https://docs.victoriametrics.com/vmagent) added `-maxLabelsPerTimeseries` and `-maxLabelValueLen` flags, which limits amount of labels and label value length for pushed to/scraped by vmagent data. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928).
```
Comment (hagen1778) [inline docs/changelog/CHANGELOG.md:26]:
```suggestion
* FEATURE: [vmagent](https://docs.victoriametrics.com/vmagent): support `-maxLabelsPerTimeseries` and `-maxLabelValueLen` flags to limit amount of labels and label value length for pushed or scraped series. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928).
```

File: docs/changelog/CHANGELOG.md [hunk 18]
```diff
@@ -23,11 +23,13 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
...
 * BUGFIX: [vmui](https://docs.victoriametrics.com/#vmui): fix rendering of isolated data points on the graph that are not connected to other points.
 * BUGFIX: [vmalert](https://docs.victoriametrics.com/vmalert): improve the correctness of alert [state restoration](https://docs.victoriametrics.com/vmalert/#alerts-state-on-restarts). Previously, it could result in false-positive alerts if alert was resolved shortly before vmalert restart.
+* BUGFIX: [vmsingle](https://docs.victoriametrics.com/single-server-victoriametrics/) changed a meaning of `-maxLabelsPerTimeseries` and `-maxLabelValueLen`. Previously excessive labels, label names and values were truncated. To prevent having corrupted data in a storage, series, which are hitting these limits will be dropped. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928)
```
Comment (hagen1778) [inline docs/changelog/CHANGELOG.md:32]: This should be a feature, because we won't port it to LTS versions of VM.
This should be also mentioned in Update notes, as it is a breaking change.

cc @f41gh7

File: dashboards/vm/victoriametrics-cluster.json [hunk 19]
```diff
@@ -3675,12 +3675,12 @@
                 "uid": "$ds"
               },
               "exemplar": true,
-              "expr": "sum(increase(vm_metrics_with_dropped_labels_total{job=~\"$job_insert\", instance=~\"$instance\"}[$__rate_interval]))",
+              "expr": "sum(increase(vm_series_dropped_total{job=~\"$job_insert\", instance=~\"$instance\"}[$__rate_interval])) by (reason)",
```
Comment (hagen1778) [inline dashboards/vm/victoriametrics-cluster.json:3678]: If we use `vm_rows_ignored_total` metric name instead, we can drop this panel

File: dashboards/vm/victoriametrics.json [hunk 20]
```diff
@@ -4191,12 +4191,12 @@
                 "uid": "$ds"
               },
               "exemplar": false,
-              "expr": "sum(increase(vm_metrics_with_dropped_labels_total{job=~\"$job\", instance=~\"$instance\"}[$__rate_interval])) by (instance)",
+              "expr": "sum(increase(vm_series_dropped_total{job=~\"$job\", instance=~\"$instance\"}[$__rate_interval])) by (instance, reason)",
```
Comment (hagen1778) [inline dashboards/vm/victoriametrics.json:4194]: If we use `vm_rows_ignored_total` metric name instead, we can drop this panel

**Review by hagen1778 — CHANGES_REQUESTED** [review]
Looks good overall!
Removing unrelated changes for `prompbmarshal`, updating error messages and metric names and we're good to go.

One more thing, @AndrewChubatiuk do you think we could update the storage limiter to accept the limits instead? I'm asking in relation to pososibility for adding labels limit to the scrape job config.

File: lib/storage/limits/limits.go (hunk 9 above)
Comment (AndrewChubatiuk) [reply lib/storage/limits/limits.go]: updated

File: app/victoria-metrics/self_scraper.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/victoria-metrics/self_scraper.go]: rebased this branch

File: lib/storage/metric_name.go [hunk 21]
```diff
@@ -511,19 +471,7 @@ func MarshalMetricNameRaw(dst []byte, labels []prompbmarshal.Label) []byte {
 	dstLen := len(dst)
 	dstSize := dstLen
 	for i := range labels {
-		if i >= maxLabelsPerTimeseries {
```
Comment (hagen1778) [inline lib/storage/metric_name.go:514]: Looks like moving this check to every ingestion protocol has the following downsides:
1. Performance impact, as every ingestion protocol will have to loop through all labels to check the limits. And then loop again to unmarshal to canonical metric name
2. Maintainance impact. Now limit check needs to be called explicitly in every protocol. It is easy to forget to do so and limits won't be working.

I think we should keep checks here for reasons above. And introduce a new function that can be excpliticly called by vmagent to check the limits.

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: lib/storage/labels_limits.go [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+			label.Name, label.Value, prompbmarshal.LabelsToString(labels), len(label.Value), maxLabelValueLen)
+	default:
+	}
+}
+
+func trackIgnoredSeriesWithTooLongLabelName(l *prompbmarshal.Label, labels []prompbmarshal.Label) {
+	ignoredSeriesWithTooLongLabelName.Add(1)
+	select {
+	case <-ignoredSeriesWithTooLongLabelNameLogTicker.C:
+		label := *l
```
Comment (hagen1778) [inline lib/storage/labels_limits.go:94]: what's the reason for doing this?

File: lib/storage/metric_name.go [hunk 23]
```diff
@@ -467,62 +463,27 @@ func (mn *MetricName) Unmarshal(src []byte) error {
...
-	if maxLabels > 0 {
-		maxLabelsPerTimeseries = maxLabels
-	}
-}
-
 // MarshalMetricNameRaw marshals labels to dst and returns the result.
 //
 // The result must be unmarshaled with MetricName.UnmarshalRaw
 func MarshalMetricNameRaw(dst []byte, labels []prompbmarshal.Label) []byte {
+	if len(labels) > maxLabelsPerTimeseries {
```
Comment (hagen1778) [inline lib/storage/metric_name.go:470]: It would be great to cover this public function with a unit test.

File: dashboards/victoriametrics-cluster.json [hunk 24]
```diff
@@ -3581,111 +3581,6 @@
           "title": "Assisted merges ($instance)",
           "type": "timeseries"
         },
-        {
```
Comment (hagen1778) [inline dashboards/victoriametrics-cluster.json:3584]: It makes sense to port `Rows ignored for last 1h` panel to vmagent, since it can ignore series now.

File: deployment/docker/rules/alerts.yml [hunk 25]
```diff
@@ -119,15 +119,3 @@ groups:
           description: "High rate of slow inserts on \"{{ $labels.instance }}\" may be a sign of resource exhaustion
             for the current load. It is likely more RAM is needed for optimal handling of the current number of active time series.
             See also https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3976#issuecomment-1476883183"
-
-      - alert: LabelsLimitExceededOnIngestion
```
Comment (hagen1778) [inline deployment/docker/rules/alerts.yml:123]: I think it makes sense to move `RowsRejectedOnIngestion` alert from cluster/single rules file to `alerts-health` file. Don't forget to rm `dashboard` annotation and update description to remove `VM` mentioning - it should be abstract which component is rejecting data.

File: app/vminsert/main.go [hunk 26]
```diff
@@ -67,8 +67,8 @@ var (
...
 	configAuthKey          = flagutil.NewPassword("configAuthKey", "Authorization key for accessing /config page. It must be passed via authKey query arg. It overrides -httpAuth.*")
 	reloadAuthKey          = flagutil.NewPassword("reloadAuthKey", "Auth key for /-/reload http endpoint. It must be passed via authKey query arg. It overrides httpAuth.* settings.")
-	maxLabelsPerTimeseries = flag.Int("maxLabelsPerTimeseries", 30, "The maximum number of labels accepted per time series. Superfluous labels are dropped. In this case the vm_metrics_with_dropped_labels_total metric at /metrics page is incremented")
-	maxLabelValueLen       = flag.Int("maxLabelValueLen", 4*1024, "The maximum length of label values in the accepted time series. Longer label values are truncated. In this case the vm_too_long_label_values_total metric at /metrics page is incremented")
+	maxLabelsPerTimeseries = flag.Int("maxLabelsPerTimeseries", 40, "The maximum number of labels per time series to be accepted. Timeseries with superfluous labels are ignored. In this case the vm_rows_ignored_total{reason=\"too_many_labels\"} metric at /metrics page is incremented")
```
Comment (hagen1778) [inline app/vminsert/main.go:70]: `timeseries` => `time series`. We should be consistent with terms.
When in doubt, refer to naming in https://docs.victoriametrics.com/keyconcepts/

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: app/vmagent/remotewrite/remotewrite.go [hunk 27]
```diff
@@ -472,6 +471,13 @@ func tryPush(at *auth.Token, wr *prompbmarshal.WriteRequest, forceDropSamplesOnF
 			rowsCountAfterRelabel := getRowsCount(tssBlock)
 			rowsDroppedByGlobalRelabel.Add(rowsCountBeforeRelabel - rowsCountAfterRelabel)
 		}
+		tmpBlock := tssBlock[:0]
+		for _, ts := range tssBlock {
+			if !storage.ExceedingLabels(ts.Labels) {
+				tmpBlock = append(tmpBlock, ts)
+			}
+		}
+		tssBlock = tmpBlock
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply app/vmagent/remotewrite/remotewrite.go:480]: do you want to use `0` to disable limits on insert as well?

[the remark this answers is not in the document]
Comment (hagen1778) [reply app/vmagent/remotewrite/remotewrite.go:480]: Yes, this sounds reasonable to me

File: lib/storage/labels_limits.go (hunk 22 above)
Comment (f41gh7) [reply lib/storage/labels_limits.go:95]: It also looks strange to me.

File: app/vmagent/remotewrite/remotewrite.go [hunk 28]
```diff
@@ -386,8 +385,8 @@ func Stop() {
 // PushDropSamplesOnFailure drops wr samples if they cannot be sent to -remoteWrite.url by any reason.
 //
 // PushDropSamplesOnFailure can modify wr contents.
-func PushDropSamplesOnFailure(at *auth.Token, wr *prompbmarshal.WriteRequest) {
-	_ = tryPush(at, wr, true)
+func PushDropSamplesOnFailure(at *auth.Token, wr *prompbmarshal.WriteRequest, labelsLimits func([]prompbmarshal.Label) bool) {
```
Comment (f41gh7) [inline app/vmagent/remotewrite/remotewrite.go:388]: This API doesn't look good to me. It's better to keep it as is and use global check ` func tryPush`.

Comment (AndrewChubatiuk) [reply app/vmagent/remotewrite/remotewrite.go:388]: don't know how to override maxLabelsPerTimeseries with a scrape config's `labels_limit` value

Comment (f41gh7) [reply app/vmagent/remotewrite/remotewrite.go:388]: WDYT? It should be applied at the same place as other scraping limits at `scrapeInternal` function. 

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/scrapework.go#L418

Comment (f41gh7) [reply app/vmagent/remotewrite/remotewrite.go:388]: Also, it should NOT be the part of this pull request.

Comment (AndrewChubatiuk) [reply app/vmagent/remotewrite/remotewrite.go]: > It should be applied at the same place as other scraping limits at scrapeInternal function.

labels_limit should override maxLabelsPerTimeseries flag value, so at least tryPush should accept labelsLimit function, where PushDropSamplesOnFailure should pass nil

File: dashboards/vmagent.json [hunk 29]
```diff
@@ -3691,6 +3691,108 @@
           ],
           "title": "Restarts ($job)",
           "type": "timeseries"
+        },
```
Comment (hagen1778) [inline dashboards/vmagent.json:3694]: this dashboard isn't synced to `vm` type

File: dashboards/vmagent.json [hunk 30]
```diff
@@ -3691,6 +3691,108 @@
...
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "$ds"
+          },
+          "description": "Shows how many rows were ignored on insertion due to corrupted or out of retention timestamps.",
```
Comment (hagen1778) [inline dashboards/vmagent.json:3700]:
```suggestion
          "description": "Shows how many samples were ignored due to various reasons. See the reason of rejection application logs.",
```

Comment (hagen1778) [reply dashboards/vmagent.json:3700]: "corrupted or out of retention timestamps" this part needs to updated in cluster version

the whole comment needs to be updated in single-node version, as it could scrape and receive writes as well

File: docs/changelog/CHANGELOG.md [hunk 31]
```diff
@@ -18,8 +18,13 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
 
 ## tip
 
+**Update note 1: changed a meaning of `-maxLabelsPerTimeseries`, `-maxLabelNameLen` and `-maxLabelValueLen`. Previously excessive labels, label names and values were truncated. With this change all series that are hitting the limits will be ignored instead.**
+
 * SECURITY: upgrade Go builder from Go1.23.3 to Go1.23.4. See the list of issues addressed in [Go1.23.4](https://github.com/golang/go/issues?q=milestone%3AGo1.23.4+label%3ACherryPickApproved).
 
+* FEATURE: [vmagent](https://docs.victoriametrics.com/vmagent): support `-maxLabelsPerTimeseries`, `-maxLabelNameLen` and `-maxLabelValueLen` flags to limit amount of labels, label name and value length for pushed or scraped series. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928).
+* FEATURE: [vmsingle](https://docs.victoriametrics.com/single-server-victoriametrics/) changed a meaning of `-maxLabelsPerTimeseries`, `-maxLabelNameLen` and `-maxLabelValueLen`. Previously excessive labels, label names and values were truncated. With this change all series, which are hitting these limits will be ignored. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6928)
```
Comment (hagen1778) [inline docs/changelog/CHANGELOG.md:26]: `maxLabelNameLen` is a new flag. Shall we mention it explicitly in the note?

**Review by hagen1778 — APPROVED** [review]
LGTM. Few comments though. I'd also appreciate have a few tests verifying it works.

File: dashboards/vmagent.json (hunk 29 above)
Comment (AndrewChubatiuk) [reply dashboards/vmagent.json:3694]: what do you mean?

File: docs/changelog/CHANGELOG.md (hunk 31 above)
Comment (AndrewChubatiuk) [reply docs/changelog/CHANGELOG.md]: updated

File: dashboards/vmagent.json (hunk 30 above)
Comment (AndrewChubatiuk) [reply dashboards/vmagent.json]: done

File: dashboards/vmagent.json (hunk 29 above)
Comment (hagen1778) [reply dashboards/vmagent.json:3694]: I think `make dashboards-sync` wasn't called.

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!
